libraries/chain/controller.cpp

```diff
@@ -1799,18 +1802,9 @@ struct controller_impl {
          EOS_ASSERT( producer_block_id == ab._id, block_validate_exception, "Block ID does not match",
                      ("producer_block_id",producer_block_id)("validator_block_id",ab._id) );
 
-         auto bsp = std::make_shared<block_state>(
-                        std::move( ab._pending_block_header_state ),
-                        b,
-                        std::move( ab._trx_metas ),
-                        protocol_features.get_protocol_feature_set(),
-                        []( block_timestamp_type timestamp,
-                            const flat_set<digest_type>& cur_features,
-                            const vector<digest_type>& new_features )
-                        {}, // validation of any new protocol features should have already occurred prior to apply_block
-                        true // signature should have already been verified (assuming untrusted) prior to apply_block
-                    );
-
+         // create completed block with existing block_state along with newly created transaction_metadata as
+         // the newly created may contain recovered keys
+         bsp->trxs = std::move( ab._trx_metas );
```

Comment:
```cpp
if( !existing_trxs_metas && !skip_auth_checks ) {
   bsp->trxs = std::move( ab._trx_metas );
}
```

Only modify the `trxs` field if the `transaction_metadata`s have the keys recovered from the signatures.

Unfortunately, this would create the same potential thread safety issues as mutating `bsp->trxs` did above. Currently, no plugins actually use the `trxs` field. Maybe we should document that it should not be used by plugins since it is not thread safe? Or come up with a better solution. Although I really prefer not having to construct a new `block_state` here as was done before because removing it really cleans up the code nicely.

Reply:
I first added an indication on `block_state` of the state of the `cached_trxs` but then realized it was really only used in `apply_block` and I think what I pushed is safer as the indicator in `block_state` can't be out of sync with `cached_trxs` now.

Reply:
Added tracking of public keys were recovered to `block_state` so we do not rely on every `transaction` having a signature.